File: examples/language-features/VbNetNet10Latest/Vb14/StringInterpolation/StringInterpolation.vb
Option Strict On

Namespace Vb14.StringInterpolation
    Public Module Interpolation
        ' An interpolated string is compiled into a formatting call. VB uses the
        ' same $"" syntax C# 6.0 introduced, in the same release wave.
        Public Function Simple(name As String, count As Integer) As String
            Return $"{name} has {count} items"
        End Function

        ' A hole may carry a format specifier after a colon...
        Public Function WithFormat(value As Double) As String
            Return $"{value:F2}"
        End Function

        ' ...and an alignment before it. Negative pads on the right.
        Public Function WithAlignment(label As String, value As Integer) As String
            Return $"{label,-10}|{value,5}"
        End Function

        Public Function WithExpression(left As Integer, right As Integer) As String
            Return $"{left} + {right} = {left + right}"
        End Function

        ' Doubling a brace escapes it.
        Public Function EscapedBraces(value As Integer) As String
            Return $"{{{value}}}"
        End Function

        ' VB has no verbatim-string prefix, because its literals never process
        ' backslash escapes in the first place — a backslash is always literal.
        Public Function BackslashIsLiteral(name As String) As String
            Return $"C:\logs\{name}.txt"
        End Function
    End Module
End Namespace

File: examples/language-features/CSharp/dotNetFramework/v4.8/CSharp_v8.0/CSharp6/StringInterpolation/StringInterpolation.cs
namespace Net48_CSharp8_Library.CSharp6.StringInterpolation
{
    public class Interpolation
    {
        // An interpolated string is a compile-time syntax for building a formatted
        // string. Each {hole} holds a real expression, checked by the compiler —
        // unlike a "{0}" placeholder whose mismatch with arguments surfaces at run time.
        public static string Simple(string name, int count)
        {
            return $"{name} has {count} items";
        }

        // A hole may carry a format specifier after a colon.
        public static string WithFormat(double value)
        {
            return $"{value:F2}";
        }

        // ...and an alignment before it. Negative pads on the right.
        public static string WithAlignment(string label, int value)
        {
            return $"{label,-10}|{value,5}";
        }

        // A hole holds any expression, not just a variable.
        public static string WithExpression(int left, int right)
        {
            return $"{left} + {right} = {left + right}";
        }

        // Doubling a brace escapes it, exactly as in a format string.
        public static string EscapedBraces(int value)
        {
            return $"{{{value}}}";
        }

        // A conditional inside a hole needs parentheses, because the colon
        // would otherwise start a format specifier.
        public static string Pluralized(int count)
        {
            return $"{count} item{(count == 1 ? string.Empty : "s")}";
        }
    }
}
